backends/platform/android/org/scummvm/scummvm/ScummVMActivity.java

```diff
@@ -2293,245 +2132,98 @@ private void copyAssetsToInternalMemory(boolean sideUpgrade) {
-			                | Intent.FLAG_GRANT_PERSISTABLE_URI_PERMISSION
-			                | Intent.FLAG_GRANT_PREFIX_URI_PERMISSION
-			);
-			startActivityForResult(intent, REQUEST_SAF);
-		}
-	}
-
-	/**
-	 * Get storage access framework tree uri. The user must have granted access via requestStorageAccessFramework
-	 *
-	 * @return Uri or null if not granted yet
-	 */
-	public Uri getStorageAccessFrameworkTreeUri() {
-		SharedPreferences sharedPref = getPreferences(Context.MODE_PRIVATE);
-		String treeStr = sharedPref.getString(getString(R.string.preference_saf_tree_key), null);
-
-		if (!TextUtils.isEmpty(treeStr)) {
-			try {
-				Log.d(ScummVM.LOG_TAG, "getStorageAccessFrameworkTreeUri: " + treeStr);
-				return Uri.parse(treeStr);
-			} catch (Exception ignored) {
+		synchronized(safSyncObject) {
+			safRequestCode = requestCode;
+			safResultCode = resultCode;
+			safResultURI = null;
+			if (resultData != null) {
+				safResultURI = resultData.getData();
 			}
+			safSyncObject.notifyAll();
 		}
-		return null;
 	}
 
-	// A method to revoke SAF granted stored permissions
-	// TODO We need a button or setting to trigger this on user's demand
-	public void clearStorageAccessFrameworkTreeUri() {
-		if (Build.VERSION.SDK_INT <= Build.VERSION_CODES.KITKAT) {
-			return;
+	// From: https://developer.android.com/training/data-storage/shared/documents-files
+	// Caution: If you iterate through a large number of files within the directory that's accessed using ACTION_OPEN_DOCUMENT_TREE, your app's performance might be reduced.
+	// Access restrictions
+	// On Android 11 (API level 30) and higher, you cannot use the ACTION_OPEN_DOCUMENT_TREE intent action to request access to the following directories:
+	// - The root directory of the internal storage volume.
+	// - The root directory of each SD card volume that the device manufacturer considers to be reliable, regardless of whether the card is emulated or removable. A reliable volume is one that an app can successfully access most of the time.
+	// - The Download directory.
+	// Furthermore, on Android 11 (API level 30) and higher, you cannot use the ACTION_OPEN_DOCUMENT_TREE intent action to request that the user select individual files from the following directories:
+	// - The Android/data/ directory and all subdirectories.
+	// - The Android/obb/ directory and all subdirectories.
+	public Uri selectWithNativeUI(boolean folder, boolean write, Uri initialURI, String prompt) {
+		// Choose a directory using the system's folder picker.
+		Intent intent;
+		if (folder) {
+			intent = new Intent(Intent.ACTION_OPEN_DOCUMENT_TREE);
+		} else {
+			intent = new Intent(Intent.ACTION_OPEN_DOCUMENT);
+			intent.addCategory(Intent.CATEGORY_OPENABLE);
+			intent.setType("*/*");
 		}
-
-		Uri treeUri;
-		if ((treeUri = getStorageAccessFrameworkTreeUri()) == null) {
-			return;
+		intent.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION | Intent.FLAG_GRANT_WRITE_URI_PERMISSION);
```

Comment:
We no longer request intent `FLAG_GRANT_PERSISTABLE_URI_PERMISSION`. According to the [intent documentation](https://developer.android.com/reference/android/content/Intent#FLAG_GRANT_PERSISTABLE_URI_PERMISSION), "_When combined with [FLAG_GRANT_READ_URI_PERMISSION](https://developer.android.com/reference/android/content/Intent#FLAG_GRANT_READ_URI_PERMISSION) and/or [FLAG_GRANT_WRITE_URI_PERMISSION](https://developer.android.com/reference/android/content/Intent#FLAG_GRANT_WRITE_URI_PERMISSION), the URI permission grant can be persisted across device reboots until explicitly revoked with [Context#revokeUriPermission(Uri, int)](https://developer.android.com/reference/android/content/Context#revokeUriPermission(android.net.Uri,%20int))_"

Is this not needed anymore? Should it be a TODO?

Reply:
After taking a look at documentation, Android source code and samples on the net, I believe these flags are useless in this way.
These flags are set in the result intent we get in the `onActivityResult` function and are not used when querying.